This directory contains the ripser-specific support functions that we
are using to explore the application of topological homology to big data
analysis. I have done two efforst in this area:

1) Computation of barcodes for the GreenGenes data base. This is done
witha combionation of C programs, ripser, and some plotting tools kept
in the Tools directory.

2) Application of barcodes to machine learning, specifically slayer, a
special input layer for persistent homology to do classification of
shapes. There is a data generator for this work in the MachineLearning
directory.


Directory Structure

ripser:

Contains a heavily cleaned up and commented version of the ripser
program. We are looking into ways to speed it up even more, both
algorithmically and computationally.

GreenGenes:

Contains the data base downloadable from :

http://greengenes.lbl.gov/Download/Sequence_Data/Fasta_data_files/).

It also contains a C program computeLowerDistanceMatrix.c that parses the
.fasta file and computes hamming distances for each of the alignments
and stores them in lower distance matrix format in a text file.

Tools:

Contains a bunch of tools that I used for my initial investigations
into ripser and barcodes on the GreenGenes data base.

Output:

Where we stage output of the different programs, specifically
createLowerDistanceMatrix and ripser itself.

MachineLearning:

Contains the makeTrainingSet.py program that makes fields of points
sampled along certain topologically interesting surfaces and stores
those models in the Output area. Also tools for plotting those
samplings.

WE PRESUME THAT THE FOLLOWING CONVENTIONS ARE ADHERED TO

To prevent loading data files into the repo, the following suffixes
have been .gitignored. Please use them for all data files

.fasta - the files provided by GreenGenes
.ldm - the files generated by createLowerDistanceMatrix or makeTrainingSet.py
.ldmf - a flattened lower distance matrix created by makeTrainingSet
.ripser - the output of ripser

Papers:

Contains relevant papers about topological homology

Notes:

Contains notes taken at various meetings

Running the thing

GreenGenes work

From the shell

If one was going to do the whole data base, something like ...

createLowerDistanceMatrix \
GreenGenes/isolated_named_strains_gg16S_aligned.fasta \
Output/isolated_named_strains_gg16S_aligned.ldm

ripser/ripser \
Output/isolated_named_strains_gg16S_aligned.ldm > \
Output/isolated_named_strains_gg16S_aligned.ripser

python \
Tools/showBarCodes.py \
Output/isolated_named_strains_gg16S_aligned.ripser

However as currently written, createLowerDistanceMatrix would take on
the order of a week to run on the whole data set, ripser would crash
for memory reasons, and I can't imagine what showBarCodes would do. So
I usually ^C the createLowerDistanceMatrix after (a while), then use
head to create a smaller ldm then use ripser/showBarCodes on that.

head -n 1024 \
Output/isolated_named_strains_gg16S_aligned.ldm \
Output/isolated_named_strains_gg16S_aligned.1024.ldm

ripser/ripser \
Output/isolated_named_strains_gg16S_aligned.1024.ldm > \
Output/isolated_named_strains_gg16S_aligned.1024.ripser

python Tools/showBarCodes.py \
Output/isolated_named_strains_gg16S_aligned.1024.ripser

Machine Learning

To run the training data generator, move into the MachineLearning folder and invoke the program

python makeTrainingSet.py ../Output/FiveClass 5

where ../Output/FiveClass is the directory where it will put the
samplings and labels, and 5 is the number of classes.

In order to feed slayer, one then has to run ripser on those things
and generate proper outputs files. This is done by invoking

generateRipserFiles ../Output/FiveClass

Be advised, this takes a long time to run generally.